lib/testing.dart

```diff
@@ -0,0 +1,133 @@
+import 'dart:async';
+
+import 'package:web_http/web_http.dart';
+
+/// A signature for callback which is invoked whenever the [TestHttpClient]
+/// recieves a request.
+typedef FutureOr<TestHttpResponse> TestClientCallback(String url, String method,
+    {String responseType,
+    String body,
+    Map<String, String> headers,
+    int timeout,
+    bool withCredentials});
+
+/// A testing implementation of [HttpClient].
+/// 
+/// The exact responses returned can be controlled with the provided 
+/// [callback].
+class TestHttpClient implements HttpClient {
+  final TestClientCallback callback;
+
+  /// Creates a new [TestHttpClient].
+  TestHttpClient(this.callback) {
+    _throwOnProductionMode();
+  }
+
+  @override
+  Stream<String> get(String url,
+      {Map<String, String> headers, int timeout, bool withCredentials}) {
+    return request(url, 'GET',
+            headers: headers,
+            timeout: timeout,
+            withCredentials: withCredentials)
+        .map((HttpResponse response) => response.body as String);
+  }
+
+  @override
+  Stream<String> post(String url, String body,
+      {Map<String, String> headers, int timeout, bool withCredentials}) {
+    return request(url, 'POST',
+            body: body,
+            headers: headers,
+            timeout: timeout,
+            withCredentials: withCredentials)
+        .map((HttpResponse response) => response.body as String);
+  }
+
+  @override
+  Stream<HttpResponse> request(String url, String method,
+      {String responseType,
+      String body,
+      Map<String, String> headers,
+      int timeout,
+      bool withCredentials}) {
+    StreamController controller;
+    controller = new StreamController(
```

Comment:
nit: collapse declaration and definition into same line:

```dart
final controller = new StreamController(
```

Reply:
This is a trick to allow me to use `controller` in the `onListen` callback - otherwise, it hasn't been declared yet.

Reply:
Oh right!